File: CPP/Bracket2.cpp
#include <cstdio>
#include <cstring>
#include <stack>
using namespace std;
void __fpurge() {
	while (getchar() != '\n');
}

bool isWellMatched(char *brackets) {
	bool isWellMatched = false;
	stack<char> bracketsBasket;
	int size = (int)strlen(brackets) - 1;
	for (int i = 0; i < size; i++) {
		if (brackets[i] == '(' || brackets[i] == '{' || brackets[i] == '[') bracketsBasket.push(brackets[i]);
		else {
			if (bracketsBasket.empty()) return isWellMatched;
			if (brackets[i] == ')' && bracketsBasket.top() == '(') bracketsBasket.pop();
			else if (brackets[i] == ']' && bracketsBasket.top() == '[') bracketsBasket.pop();
			else if (brackets[i] == '}' && bracketsBasket.top() == '{') bracketsBasket.pop();
		}
	}
	if (bracketsBasket.empty() && size > 0) isWellMatched = true;
	return isWellMatched;
}

int main() {
	int cases;
	scanf("%d", &cases); __fpurge();
	while (cases--) {
		char brackets[1024];
		fgets(brackets, 1024, stdin);
		if (isWellMatched(brackets)) fputs("YES\n", stdout);
		else fputs("NO\n", stdout);
	}
}

File: JAVA/Bracket2.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;


public class Bracket2 {

	public static void main(String[] args) {
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
			int cases = Integer.parseInt(reader.readLine());
			
			while (cases-- > 0) {
				char[] brackets = reader.readLine().toCharArray();
				if (isWellMatched(brackets)) writer.append("YES\n");
				else writer.append("NO\n");
			}
			writer.flush();
			writer.close(); reader.close();
			
		} catch (Exception e) {
			e.printStackTrace();
		} 
	}
	
	public static boolean isWellMatched(char[] brackets) {
		boolean isWellMatched = false;
		Stack<Character> braketsBasket = new Stack<Character>();
		for (int i = 0; i < brackets.length; i++) {
			switch (brackets[i]) {
			case '(' : case '[' : case '{' : braketsBasket.push(brackets[i]); break;
			case ')' : if (!braketsBasket.isEmpty() && braketsBasket.peek() == '(') braketsBasket.pop(); else return isWellMatched; break;
			case ']' : if (!braketsBasket.isEmpty() && braketsBasket.peek() == '[') braketsBasket.pop(); else return isWellMatched; break;
			case '}' : if (!braketsBasket.isEmpty() && braketsBasket.peek() == '{') braketsBasket.pop(); else return isWellMatched;	break;
			}
		}
		if (braketsBasket.isEmpty()) isWellMatched = true;
		return isWellMatched;
	}

}
